### PR TITLE
Port upstream PR #1428: Client Mode (Empty) — multitenancy hardening

### PR DESCRIPTION
**Pull request by krukow**
## Summary

Ports upstream [github/copilot-sdk PR #1428](https://github.com/github/copilot-sdk/pull/1428) — "Multitenancy hardening: Client Mode (Empty)" — that was deferred from round 6 ([copilot-community-sdk/copilot-sdk-clojure#113](https://github.com/copilot-community-sdk/copilot-sdk-clojure/pull/113)). This is the dedicated follow-up sync round for that one PR.

Adds the `:mode #{:copilot-cli :empty}` client option (default `:copilot-cli` — historical behavior preserved). `:empty` mode is the hardened multitenancy posture for SaaS hosts that run sessions on behalf of multiple users and must isolate sessions from local machine state.

This is a **non-breaking** addition. `:copilot-cli` mode is the default, and the only behavioral change for that mode is that `:tool-filter-precedence "excluded"` is now always emitted on `session.create` / `session.resume` so the ordering between `:available-tools` and `:excluded-tools` is deterministic across CLI versions. Bare `"*"` in tool filters is rejected at the SDK boundary in both modes (mirrors upstream `resolveToolFilterOptions`).

Session plan: `~/.copilot/session-state/b794e17b-2c5a-42ed-90fe-63b823312966/plan.md` (local; key decisions captured in this PR description and the per-step checkpoints).

## What `:empty` mode does

| Scope | Behavior |
|---|---|
| Constructor | Requires ONE of `:copilot-home`, `:session-fs`, `:cli-url`, or `:is-child-process?` so the CLI never falls back to the user's home directory |
| Constructor | Forces `COPILOT_DISABLE_KEYTAR=1` on the spawned CLI (via the `cli-env-overrides :overrides` slot — caller cannot override) |
| Session validation | Every `create-session` / `resume-session` (sync + async) must supply `:available-tools` (`[]` is legitimate — the key just has to be present so silently-empty filters cannot happen) |
| Session config | Spreads 9 safe defaults UNDER caller config (caller always wins): `:enable-session-telemetry? false`, `:mcp-oauth-token-storage :in-memory`, `:skip-embedding-retrieval true`, `:embedding-cache-storage :in-memory`, `:enable-on-demand-instruction-discovery false`, `:enable-file-hooks false`, `:enable-host-git-operations false`, `:enable-session-store false`, `:enable-skills false` |
| System message | Strips `environment_context` unless the app has taken control of it: no `:system-message` → `{:mode customize :sections {:environment_context {:action remove}}}`; `:append` is promoted to `:customize` preserving content + env-context removed; `:customize` without env-context override gets one added; `:customize` with explicit env-context override is left untouched; `:replace` passes through unchanged |
| Post-create RPC | After `session.create` / `session.resume` succeeds, issues a follow-up `session.options.update` with four overridable feature flags (`:skip-custom-instructions true`, `:custom-agents-local-only true`, `:coauthor-enabled false`, `:manage-schedule-enabled false`) plus `:installed-plugins []`. On RPC failure the SDK disconnects + removes the half-configured session before rethrowing |

In `:copilot-cli` mode the `session.options.update` RPC is only sent when the caller explicitly set one of the four new flags; if the resulting patch is empty the RPC is skipped entirely.

## New public API surface

- **Client option**: `:mode #{:copilot-cli :empty}` (default `:copilot-cli`).
- **Session-config keys** (forwarded via `session.options.update`, NOT `session.create`):
  - `:skip-custom-instructions` (boolean)
  - `:custom-agents-local-only` (boolean)
  - `:coauthor-enabled` (boolean)
  - `:manage-schedule-enabled` (boolean)
- **New namespace `github.copilot-sdk.tool-set`**: source-qualified tool filter constructors (`builtin`/`mcp`/`custom`/`builtins`) plus the `isolated-builtins` / `isolated` preset (parity with upstream `BuiltInTools.Isolated`). Bare `"*"` rejected at construction time.

See [`doc/reference/API.md` → Client Mode (Empty)](https://github.com/copilot-community-sdk/copilot-sdk-clojure/blob/upstream-sync/pr-1428-client-mode-empty/doc/reference/API.md#client-mode-empty) and [Tool Sets](https://github.com/copilot-community-sdk/copilot-sdk-clojure/blob/upstream-sync/pr-1428-client-mode-empty/doc/reference/API.md#tool-sets) for full reference docs.

### Why one mode flag and not nine

Upstream PR #1428 introduces a single `mode` enum so multi-tenant hosts can opt into the entire hardened posture (env-var override, config defaults, system-message normalization, options.update flags, validation) with one knob. The Clojure port preserves this — the per-feature flags exist (`:enable-skills`, `:skip-embedding-retrieval`, etc., from round 6) but `:mode :empty` is the one-line opt-in.

### `session.options.update` is a SEPARATE RPC, not a `session.create` field

The four feature flags do NOT ride on `session.create` — they ship in a follow-up `session.options.update` RPC that the SDK issues after the session is registered. This mirrors upstream and means:

- If `session.create` succeeds but `session.options.update` fails, the SDK MUST clean up (else we leak a half-configured session). `cleanup-failed-options-update!` calls both `session/disconnect!` (CLI-side cleanup via `session.destroy`) AND `session/remove-session!` (in-memory registry cleanup); both are idempotent.
- The async path (`<apply-session-options-update!`) returns a channel that yields `:ok` on success (including the no-op case) or a `Throwable` on failure (after cleanup), matching the existing async error convention.
- If the computed patch is empty (e.g. `:copilot-cli` mode + no flags set), the RPC is skipped entirely.

### `:available-tools` required, but `[]` is legitimate

`validate-empty-mode-session-requirements!` uses `(contains? config :available-tools)` — NOT `seq`. An explicit empty vector `[]` means "no tools" and is legitimate; only an absent key is rejected. This mirrors upstream's `resolveToolFilterOptions` where the distinction between "undefined" and "empty list" is meaningful.

### Always-emit `:tool-filter-precedence`

Both modes now always send `toolFilterPrecedence: "excluded"` on `session.create` and `session.resume`. This is the only behavioral change to `:copilot-cli` mode, and it makes the precedence between `:available-tools` and `:excluded-tools` deterministic regardless of CLI version (older CLIs might have defaulted differently).

### `::client-mode` not `::mode`

`::specs/mode` was already defined at line 899 for `message-options` (`#{:enqueue :immediate}`). The new spec is named `::specs/client-mode` and validated in `::client-options` via a predicate (mirrors the existing `::remote-session-mode` pattern further down).

## Validation

- `bb test` (unit + integration): **339 tests / 1578 assertions / 0 failures / 0 errors** (+ 18 new tests this round across constructor validation, env-var override, wire payload mode-defaults, system-message normalization, and `session.options.update` RPC including async path + cleanup-on-failure)
- `bb validate-docs`: clean
- `bb jar`: succeeds
- `COPILOT_E2E_TESTS=true bb test`: 2 flaky E2E timeouts (`test-e2e-blob-attachment`, `test-e2e-list-sessions`) — both pre-existing model-latency timeouts on `main`, neither exercises `:mode :empty` or any new code path. Not regressions.
- `examples/empty_mode.clj`: compile-checked. Not included in `run-all-examples.sh` because empty mode disables the local keychain and the example requires a BYOK API key (matches how `byok_provider.clj` and `mcp_local_server.clj` are handled — see [`examples/README.md`](https://github.com/copilot-community-sdk/copilot-sdk-clojure/blob/upstream-sync/pr-1428-client-mode-empty/examples/README.md#example-20-empty-multitenancy-mode-empty_modeclj)).

## Commits

1. [`feat(tool-set): add github.copilot-sdk.tool-set namespace + isolated preset`](https://github.com/copilot-community-sdk/copilot-sdk-clojure/commit/451d47c)
2. [`feat(client): implement client mode :empty (multitenancy hardening)`](https://github.com/copilot-community-sdk/copilot-sdk-clojure/commit/94ad7ab)
3. [`docs: document client mode :empty + empty_mode example`](https://github.com/copilot-community-sdk/copilot-sdk-clojure/commit/a3af6b3)

Each commit individually builds; commits 1 and 2 each pass tests at HEAD.

---

<sub><em> Generated via Copilot <a href="https://adaptivepatchwork.com/ai-attribution">on behalf</a> of @krukow</em></sub>

### CODE REVIEW

File: src/github/copilot_sdk/client.clj [hunk 1]
```diff
@@ -1420,6 +1481,43 @@
...
+        (assoc config :system-message {:mode :customize :sections ec-remove})
+
+        (= :replace (:mode sm))
+        config
+
+        (= :customize (:mode sm))
+        (if (contains? (:sections sm) :environment-context)
+          config
+          (update-in config [:system-message :sections]
+                     (fnil assoc {}) :environment-context {:action :remove}))
```
[the remark this answers is not in the document]
Comment (krukow) [reply src/github/copilot_sdk/client.clj:1514]: <sub><em> Generated via Copilot <a href="https://adaptivepatchwork.com/ai-attribution">on behalf</a> of @krukow</em></sub>

This specific claim ("will throw") is a **false positive**: in Clojure, `(contains? nil :anything)` returns `false`, not an exception (verified: `clojure -e '(contains? nil :foo)'` → `false`). The else branch then runs `(update-in config [:system-message :sections] (fnil assoc {}) :environment-context {:action :remove})` — `fnil` substitutes `{}` for the nil `:sections`, so the SDK correctly produces `{:mode :customize :sections {:environment-context {:action :remove}}}`, which is exactly the behavior you describe as the intent.

That said, the test suite did not have explicit coverage of the "no `:sections` key at all" case (the existing customize-without-env-context test passes a `:tone` section). To lock the semantics in, I added [`test-empty-mode-system-message-customize-no-sections-key`](https://github.com/copilot-community-sdk/copilot-sdk-clojure/blob/f8b7b20/test/github/copilot_sdk/integration_test.clj#L5723-L5728) in [`f8b7b20`](https://github.com/copilot-community-sdk/copilot-sdk-clojure/commit/f8b7b20) so a future refactor that swaps `contains?` for something that *does* throw on nil will be caught immediately.

File: src/github/copilot_sdk/client.clj [hunk 2]
```diff
@@ -1457,6 +1555,123 @@
...
+                      (ex-info (str "session.options.update failed: "
+                                    (get-in response [:error :message] "RPC error"))
+                               {:session-id session-id :error (:error response)}))]
+            (if err
+              (do
+                (log/warn "session.options.update failed; cleaning up session " session-id)
+                (cleanup-failed-options-update! client session-id)
+                (>! out err))
+              (>! out :ok))
+            (close! out))))
```
[the remark this answers is not in the document]
Comment (krukow) [reply src/github/copilot_sdk/client.clj:1673]: <sub><em> Generated via Copilot <a href="https://adaptivepatchwork.com/ai-attribution">on behalf</a> of @krukow</em></sub>

Valid concern — fixed in [`f8b7b20`](https://github.com/copilot-community-sdk/copilot-sdk-clojure/commit/f8b7b20). `cleanup-failed-options-update!` is now wrapped in `async/thread` and the go block `<!`s the result before delivering the Throwable:

```clojure
(if err
  (do
    (log/warn "session.options.update failed; cleaning up session " session-id)
    ;; cleanup-failed-options-update! calls blocking session.destroy
    ;; via proto/send-request! — offload to async/thread so it does
    ;; not starve the core.async dispatch threadpool.
    (<! (async/thread (cleanup-failed-options-update! client session-id)))
    (>! out err))
  (>! out :ok))
```

This matches the existing convention in `session.clj` (see [`session.clj:93-100`](https://github.com/copilot-community-sdk/copilot-sdk-clojure/blob/f8b7b20/src/github/copilot_sdk/session.clj#L93-L100) for the on-event handler offload comment: "Uses async/thread to avoid blocking core.async dispatch threads, since user handlers may perform blocking I/O").

Also added [`test-empty-mode-options-update-async-failure-cleans-up-session`](https://github.com/copilot-community-sdk/copilot-sdk-clojure/blob/f8b7b20/test/github/copilot_sdk/integration_test.clj#L5908-L5937) covering the async failure path end-to-end (Throwable yielded on the result channel + session removed from the in-memory registry).
